src/pass.jl

```diff
@@ -1,42 +1,67 @@
-using IRTools
-using IRTools: IR, Argument, isexpr
-using LinearAlgebra
-
-function pass(ir)
-  for (x, st) in ir
-    (isexpr(st.expr, :call) && !IRTools.isprimitive(ir, st.expr.args[1])) || continue
-    ir[x] = Expr(:call, Argument(1), st.expr.args...)
-  end
-  return ir
+export hoist_alloc, Buffer
+using Zygote
+
+using Cassette, LinearAlgebra
+using Cassette: @context, overdub
+
+@context BuffCtx
+
+function hoist_alloc(f, b::Buffer)
+    clear!(b)
+    return overdub(BuffCtx(metadata=b), f)
+end
+
+# Cassette.prehook(ctx::BuffCtx, f, args...) = println(typeof(f), "  ", typeof(args))
+
+@inline function Cassette.overdub(ctx::BuffCtx, ::typeof(similar), bc::Broadcast.Broadcasted{Broadcast.DefaultArrayStyle{N}}, ::Type{T}) where {T, N}
+    isbitstype(T) || return similar(bc, T)
+    return alloc(ctx.metadata, Array{T, N}, length.(Broadcast.axes(bc)))
 end
 
-@generated function (b::Buffer)(f, args...)
-  m = IRTools.meta(Tuple{f,args...})
-  m == nothing && return :(f(args...))
-  # Core.println((f, args...))
-  ir = IR(m)
-  ir = IRTools.spliceargs!(m, ir, (Symbol("#buf#"),Any))
-  ir = pass(ir)
-  ir = IRTools.varargs!(m, ir, 2)
-  IRTools.argnames!(m, Symbol("#buf#"), :f, :args)
-  return IRTools.update!(m, ir)
+# NOTE: should use inferred type of similar instead? or just do overloading
+@inline function Cassette.overdub(ctx::BuffCtx, ::typeof(similar), X::AbstractArray, ::Type{T}, dims::Dims{N}) where {T, N}
+    isbitstype(T) || return similar(X, T, dims)
+    return alloc(ctx.metadata, Array{T, N}, dims)
 end
 
-(buf::Buffer)(::Type{Array{T,N}}, ::UndefInitializer, d::Vararg{Int,N}) where {T,N} =
-  alloc(buf, Array{T,N}, d)
+@inline function Cassette.overdub(ctx::BuffCtx, ::typeof(similar), X::AbstractArray{T}, dims::Dims{N}) where {T, N}
+    isbitstype(T) || return similar(X, T, dims)
+    return alloc(ctx.metadata, Array{T, N}, dims)
+end
 
-# Some definitions we're not interested in
-# Limits compilation and avoids some perf traps
+@inline function Cassette.overdub(ctx::BuffCtx, ::typeof(copy), X::Array{T, N}) where {T, N}
+    isbitstype(T) || return copy(X)
+    new = alloc(ctx.metadata, Array{T, N}, size(X))
+    copyto!(new, X)
+    return new
+end
 
-for f in [:, getproperty, LinearAlgebra.mul!, Base.promote_op, size, Base.to_shape,
-          Broadcast.broadcasted, Broadcast.instantiate, Broadcast.preprocess,
-          Broadcast.combine_eltypes, copyto!, Broadcast.copyto_nonleaf!,
-          Broadcast.axes]
-  @eval @inline (::Buffer)(::typeof($f), a...) = $f(a...)
+@inline function Cassette.overdub(ctx::BuffCtx, ::Type{Array{T, N}}, ::UndefInitializer, d::Vararg{Int, N}) where {T, N}
+    isbitstype(T) || return Array{T, N}(undef, d)
+    T === UInt8 && return Array{T, N}(undef, d) # make Zygote stack happy
```

Comment:
This is one place I had to workaround @MikeInnes but I'm not sure where is allocating those pullback stacks?

Reply:
If there's a type for stacks, I guess I could just add that type's constructor to white list then I won't need this line.